eos/c-decays/d-to-psd-l-nu.cc

```diff
@@ -0,0 +1,647 @@
+ * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 59 Temple
+ * Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+#include <eos/c-decays/d-to-psd-l-nu.hh>
+#include <eos/form-factors/form-factors.hh>
+#include <eos/maths/integrate.hh>
+#include <eos/maths/power-of.hh>
+#include <eos/models/model.hh>
+#include <eos/utils/destringify.hh>
+#include <eos/utils/kinematic.hh>
+#include <eos/utils/options-impl.hh>
+#include <eos/utils/private_implementation_pattern-impl.hh>
+
+#include <map>
+#include <string>
+
+namespace eos
+{
+    namespace d_to_psd_l_nu
+    {
+        struct Amplitudes
+        {
+            // helicity amplitudes, cf. [DDS:2014A] eqs. 13-14
+            complex<double> h_0;
+            complex<double> h_t;
+            complex<double> h_S;
+            complex<double> h_T;
+            complex<double> h_tS;
+            double v;
+            double p;
+            double NF;
+        };
+    }
+
+    template <>
+    struct Implementation<DToPseudoscalarLeptonNeutrino>
+    {
+        std::shared_ptr<Model> model;
+
+        Parameters parameters;
+
+        SwitchOption opt_Q;
+        SwitchOption opt_q;
+        SwitchOption opt_I;
+
+        UsedParameter m_D;
+
+        UsedParameter tau_D;
+
+        UsedParameter m_P;
+
+        LeptonFlavorOption opt_l;
+
+        UsedParameter m_l;
+
+        UsedParameter g_fermi;
+
+        UsedParameter hbar;
+
+        const double isospin_factor;
+
+        UsedParameter mu;
+
+    	static const std::vector<OptionSpecification> options;
+
+        std::function<double (const double &)> m_Q_msbar;
+        std::function<complex<double> ()> v_cQ;
+        std::function<WilsonCoefficients<ChargedCurrent> (LeptonFlavor, bool)> wc;
+
+        GSL::QAGS::Config int_config;
+
+        SpecifiedOption opt_cp_conjugate;
+
+        bool cp_conjugate;
+
+        std::shared_ptr<FormFactors<PToP>> form_factors;
+
+        // { Q, q, I } -> { process, m_D, m_P, c_I }
+        // Q: d, s; the quark flavor in the weak transition
+        // q: u, d, s: the spectar quark flavor
+        // I: 1, 0, 1/2: the total isospin of the daughter meson
+        // process: string that can be used to obtain the form factor
+        // D: name of the D meson
+        // P: name of the daughter meson
+        // c_I: isospin factor by which the amplitudes are multiplied
+        static const std::map<std::tuple<char, char, std::string>, std::tuple<std::string, std::string, std::string, double>> process_map;
+
+        inline std::string _process() const
+        {
+            const char Q = opt_Q.value()[0];
+            const char q = opt_q.value()[0];
+            const std::string I = opt_I.value();
+            const auto p = process_map.find(std::make_tuple(Q, q, I));
```

Comment:
Using ``QuarkFlavourOption, this should simplify to a single line.

Reply:
I don't see what you mean here

Reply:
I meant that you can simply use ``const auto p = process_map.find(std::make_tuple(opt_Q.value(), opt_q.value(), opt_I.value()));``

Also, we should probably add ``operator QuarkFlavour()`` to a ``QuarkFlavourOption`` and similar. I'll try that out somewhere else!

Reply:
I can still apply this

Reply:
No, let's apply that to the entire tree sometime later.